homeassistant/components/zha/config_flow.py

```diff
@@ -1,80 +1,130 @@
-            test = await check_zigpy_connection(
-                user_input[CONF_USB_PATH], user_input[CONF_RADIO_TYPE], database
+            user_selection = user_input[CONF_DEVICE_PATH]
+            if user_selection == CONF_MANUAL_PATH:
+                return await self.async_step_pick_radio()
+
+            port = ports[list_of_ports.index(user_selection)]
+            dev_path = await self.hass.async_add_executor_job(
+                get_serial_by_id, port.device
             )
-            if test:
+            auto_detected_data = await detect_radios(dev_path)
+            if auto_detected_data is not None:
+                title = f"{port.description}, s/n: {port.serial_number or 'n/a'}"
+                title += f" - {port.manufacturer}" if port.manufacturer else ""
+                return self.async_create_entry(title=title, data=auto_detected_data,)
+
+            # did not detect anything
+            self._device_path = dev_path
+            return await self.async_step_pick_radio()
+
+        schema = vol.Schema({vol.Required(CONF_DEVICE_PATH): vol.In(list_of_ports)})
+        return self.async_show_form(step_id="user", data_schema=schema)
+
+    async def async_step_pick_radio(self, user_input=None):
+        """Select radio type."""
+
+        if user_input is not None:
+            self._radio_type = user_input[CONF_RADIO_TYPE]
+            return await self.async_step_port_config()
+
+        schema = {vol.Required(CONF_RADIO_TYPE): vol.In(sorted(RadioType.list()))}
+        return self.async_show_form(
+            step_id="pick_radio", data_schema=vol.Schema(schema),
+        )
+
+    async def async_step_port_config(self, user_input=None):
+        """Enter port settings specific for this type of radio."""
+        errors = {}
+        app_cls = RADIO_TYPES[self._radio_type][CONTROLLER]
+
+        if user_input is not None:
+            self._device_path = user_input.get(CONF_DEVICE_PATH)
+            if await app_cls.probe(user_input):
+                serial_by_id = await self.hass.async_add_executor_job(
+                    get_serial_by_id, user_input[CONF_DEVICE_PATH]
+                )
+                user_input[CONF_DEVICE_PATH] = serial_by_id
                 return self.async_create_entry(
-                    title=user_input[CONF_USB_PATH], data=user_input
+                    title=user_input[CONF_DEVICE_PATH],
+                    data={CONF_DEVICE: user_input, CONF_RADIO_TYPE: self._radio_type},
                 )
             errors["base"] = "cannot_connect"
 
+        schema = {
+            vol.Required(
+                CONF_DEVICE_PATH, default=self._device_path or vol.UNDEFINED
+            ): str
+        }
         return self.async_show_form(
-            step_id="user", data_schema=vol.Schema(fields), errors=errors
+            step_id="port_config",
+            data_schema=app_cls.SCHEMA_DEVICE.extend(schema),
```

Comment:
Is the `app_cls.SCHEMA_DEVICE` defined in a library? It's not good to depend on third party libraries directly for the schema as we want to control the strings to be able to translate them.

Reply:
What if they’re our libs? In that case we control them

Reply:
Our they part of the home assistant organization?

I don't think it matters since we want to guard against changes in a different api that propagates to our api without any explicit changes of our api. There needs to be a level of decoupling in between that we define here in the source.

Reply:
So I get that point but the guidance is always that supporting code for an integration needs to go into a lib on pypi. That being the case, in order for us to leverage the same features (Vol schema in this instance) in our lib that supports the integration we’d have to duplicate the schemas on both sides: in HA and in the lib that supports the integration. Wouldn’t it make more sense and be less error prone to share the code to ensure it’s consistent? I’d argue there is more value sharing it (especially since the component doesn’t work without the lib anyway) than there is otherwise. There would be less maintenance since the same schema wouldn’t have to be double maintained and you don’t have the chance that they become out of sync on accident. Does that make sense?

Reply:
We just need to define the form strings here, not the whole schema. We can define a dict that maps between our form strings and the strings of the library schema, and iterate over either the schema or our dict and create a new schema from the library schema.

Reply:
So yes, this comes from a separate lib, but we control the libs and we try to keep the schema consistent accross different radios, but sill each different radio supported by ZHA may have slightly different options for connections and this would allow handling those differences without requiring changes in HA. And at the same time presents user with the options only applicable only for that radio. 
For example: EZSP based radios mostly need serial device path and use a default `baudrate` parameter of `57600`, however Elelabs EZSP based radio needs baudrate of `115200`. So for EZSP we need to present serial path and baudrate options. 
On the other hand, deCONZ (Conbee) radios only need the serial device path. So instead of hardcoding schemas for each radio (which 95% users won't need to use) we decided to keep that in the radio lib. 
And as I mentioned in the PR description, in 95% of the time I'm not expecting user even seeing this.

Although I see your point about the localization and as a work around we just could keep aggregate of those strings updated in `strings.json`

Reply:
> We can define a dict that maps between our form strings and the strings of the library schema, and iterate over either the schema or our dict and create a new schema from the library schema

Hrm, wouldn't it be easier just to keep `strings.json` updated with the "union" of schema strings from all radio libraries? It is already covering all three radios and I'm not expecting this particular schema to be significantly different accross all supported radios. 
We could add a "mapping" and rebuild the schema, but IMO it is over-complicating things.

Reply:
We need to have a guard in the code.

Reply:
I could try just drop unknown keys from schema then

Reply:
@MartinHjelmare this honestly feels arbitrary. I don’t want this to come off as offensive because I am grateful for the time and effort you give in your reviews of our stuff but I don’t see the benefit to this. 

Are you saying that nothing in a lib should ever be directly referenced by anything translatable in HA? so anything that could potentially appear in the UI would need a defensive copy or some other form of guard? Or only in the case of vol schemas? If it’s the later, isn’t that even more arbitrary? 

You proposed having a mapping layer. Say we did that and in a subsequent PR we changed the libs and not the mappings that exist to guard HA... things still break because the mappings are no longer correct. We either have to update the translations the way it is now or we have to update the mapping code with any solution that handles mapping. What’s the difference in the code at that point and what have we actually guarded?

In the cases of new key additions or key deletions we would have more work in the mapping approach: 3 changes vs 2 because we would have to add/delete the translation, update the mapping and update the lib. How are we not just creating additional risk and work here? 

@Adminiuga i’d argue that I’d rather have untranslated keys show on the screen and everything still work rather than dropping keys and potentially breaking the integration for the sake of ensuring translation but that’s just my opinion

Reply:
If we have a guard it will protect for breakage. We will not show form keys that haven't been explicitly included. So the new features won't be active unless explicitly activated.

Reply:
ok, added guard.